src/extensions/nuget/Microsoft.DotNet.UpgradeAssistant.Extensions.NuGet/NuGetTransitiveDependencyIdentifier.cs

```diff
@@ -137,39 +151,19 @@ public TargetGraphLookup(RestoreTargetGraph graph)
 
             public bool Contains(NuGetReference key) => TryFind(key, out _);
 
-            public IEnumerator<NuGetReference> GetEnumerator()
-            {
-                foreach (var item in _graph.Flattened)
-                {
-                    var library = item.Data.Match.Library;
-
-                    yield return new(library.Name, library.Version.ToString());
-                }
-            }
+            public IEnumerator<NuGetReference> GetEnumerator() => _graph.Flattened.Select(item => new NuGetReference(item.Data.Match.Library.Name, item.Data.Match.Library.Version.ToString())).GetEnumerator();
 
             IEnumerator IEnumerable.GetEnumerator() => GetEnumerator();
 
-            IEnumerator<IGrouping<NuGetReference, NuGetReference>> IEnumerable<IGrouping<NuGetReference, NuGetReference>>.GetEnumerator()
-            {
-                foreach (var item in _graph.Flattened)
-                {
-                    yield return new ResolveResultGrouping(item.Data);
-                }
-            }
+            IEnumerator<IGrouping<NuGetReference, NuGetReference>> IEnumerable<IGrouping<NuGetReference, NuGetReference>>.GetEnumerator() => _graph.Flattened.Select(item => new ResolveResultGrouping(item.Data)).GetEnumerator();
 
             private bool TryFind(NuGetReference package, [MaybeNullWhen(false)] out ResolveResultGrouping result)
             {
-                foreach (var item in _graph.Flattened)
-                {
-                    if (string.Equals(item.Data.Match.Library.Name, package.Name, StringComparison.OrdinalIgnoreCase))
-                    {
-                        result = new(item.Data);
-                        return true;
-                    }
-                }
+                var foundItem = _graph.Flattened.FirstOrDefault(item => string.Equals(item.Data.Match.Library.Name, package.Name, StringComparison.OrdinalIgnoreCase));
```

Comment:
The non-linq version likely allocate less. If this is a hot path, you may save some GC time.